### PR TITLE
andreasbuhr-cppcoro: add lib in package_info()

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **lib/1.0**

@hnOsmium0001 Could you provide an example relying on symbols in the lib please?

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by hnOsmium0001** [comment]
Here is the patch that makes the test package rely on lib being linked. Should be able to just `git apply` it.
```patch
diff --git a/recipes/andreasbuhr-cppcoro/all/test_package/test_package.cpp b/recipes/andreasbuhr-cppcoro/all/test_package/test_package.cpp
index cddfeea6..93b96200 100644
--- a/recipes/andreasbuhr-cppcoro/all/test_package/test_package.cpp
+++ b/recipes/andreasbuhr-cppcoro/all/test_package/test_package.cpp
@@ -1,12 +1,20 @@
 #include <vector>
 #include <iostream>
 #include <cppcoro/generator.hpp>
+#include <cppcoro/async_mutex.hpp>
+#include <cppcoro/task.hpp>
 
 cppcoro::generator<int> intYielder() {
        co_yield 0;
        co_yield 1;
 }
 
+// Compile only test, uses code from the library instead of solely in headers
+cppcoro::async_mutex mutex;
+cppcoro::task<> testMutex() {
+       cppcoro::async_mutex_lock lock = co_await mutex.scoped_lock_async();
+}
+
 int main() {
        std::vector<int> v;
        for (int n : intYielder()) {
```

**Comment by SpaceIm** [comment]
Thanks, indeed with this modification, link of test package fails if lib is not in `cpp_info.libs`

**Comment by SpaceIm** [comment]
shared lib is illformed, they should link to pthread in CMakeLists

**Comment by hnOsmium0001** [comment]
Seems like MSVC has updated their coroutine implementation. We may want to change the `/await` flag to `/await:strict` (haven't tested it yet)
https://github.com/microsoft/STL/blob/main/stl/inc/coroutine#L18

**Comment by SpaceIm** [comment]
Would it work for Visual Studio 2017? Currently only 2019 fails.

**Comment by SpaceIm** [comment]
Maybe add `/await:heapelide`?
https://github.com/andreasbuhr/cppcoro/blob/7cc9433436fe8f2482138019cfaafce8e1d7a896/cmake/FindCoroutines.cmake#L117-L126

**Comment by SpaceIm** [comment]
/await:strict is available since Visual Studio 16.10: https://docs.microsoft.com/en-us/cpp/build/reference/await-enable-coroutine-support?view=msvc-160

**Comment by hnOsmium0001** [comment]
## What actually happened
There are 2 coroutine implementations in Visual Studio 2019:
1. ABI 1, which is enabled by the /await flag
2. ABI 2, which is enabled by either /std:c++latest (/std:c++20 if this flag ever gets added) or /await:strict in VS 16.10.

I thought the CI was running 16.10, causing the ABI mismatch, but it was actually for a different cause (though changing to /await:strict coincidentally fixed it).

In [build 4](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/6197/4/andreasbuhr-cppcoro/cci.20210113//summary.json), the VS 16 tests failed because we force added /await flag, causing MSVC to use ABI 1. cppcoro's `FindCoroutine.cmake` file checked if it can compile without any flags - which it could due to c++ version set to 20, so MSVC defaulted to ABI 2. Ultimately causing a link failure with test_package.obj.

The [build 5](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/6197/5/andreasbuhr-cppcoro/cci.20210113//summary.json), the VS 16 tests succeed because by not passing the /await option, the compiler defaulted to ABI because c++ version is set to 20 in `test_package/CMakeLists.txt`.
It didn't actually know the /await:strict flag - CI is running MSVC 19.28 (aka VS 16.9), while the flag is added in MSVC 19.29 (aka VS 16.10).
```
cl : command line warning D9002: ignoring unknown option '/await:strict' [C:\J\w\BuildSingleReference@6\conan-center-index\recipes\andreasbuhr-cppcoro\all\test_package\build
```

As a side note, I think in VS 16.10, /await implements <experimental/coroutine> instead of <coroutine>, evident from the Microsoft STL link I posted earlier. Just putting this here in case it breaks in the future.

\<rant>
The visual studio people loves introducing command line breaking changes in a major version, a big pain in the ass to figure out why things broke. Maybe we should treat every minor version increase as a breaking change...
\</rant>

## The fix
There are several cases that needs to be handled if we want a perfect solution:
1. **User using Visual Studio 2015 or 2017**
   We can simply keep /await here. There is no such ABI nonsense.
2. **User using Visual Studio 2019, with c++17 or below**
   VS does support coroutines in non-c++20 mode, so we need to add /await. This results in ABI 1.
   No need to handle /await:strict (for now) because cppcoro doesn't handle that flag at all, it just goes straight to /await (note that this would result in ABI 2 instead).
3. **User using Visual Studio 2019, with c++20 or above**
   We don't have to do anything here, since MSVC enables coroutine capability for c++20 already. This results in ABI 2.

(1) is trivial. (2) and (3) is tricky because the conan recipe has no way of knowing what the user is compiler their code with. `compiler.cppstd` doesn't work because they may build the libraries with c++20, but compile their own code with c++17.
There is also the problem that older MSVCs (don't know the exact numbers) shipped with Visual Studio 2019 _requires_ adding /await for coroutines, even with c++20 enabled.

Given the situation, I think the best solution here is to just say "you can't use cppcoro without c++20 and the latest MSVC" -- just do nothing for Visual Studio 2019.

## References
https://docs.microsoft.com/en-us/cpp/build/reference/await-enable-coroutine-support?view=msvc-160
https://github.com/microsoft/STL/blob/main/stl/inc/coroutine#L13-L15
https://github.com/andreasbuhr/cppcoro/blob/master/cmake/FindCoroutines.cmake

**Comment by hnOsmium0001** [comment]
Giving it a second thought, we can just read CMakeCache.txt produced by FindCoroutine.cmake and take the selected flags there. This should meet the users' expectations while keeping our job simple.

File: recipes/andreasbuhr-cppcoro/all/conanfile.py [hunk 1]
```diff
@@ -91,18 +93,14 @@ def package(self):
...
-        comp = self.cpp_info.components["cppcoro"]
-        comp.names["cmake_find_package"] = "cppcoro"
-        comp.names["cmake_find_package_multi"] = "cppcoro"
-
+        self.cpp_info.libs = ["cppcoro"]
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            self.cpp_info.system_libs.append("pthread")
         if self.settings.compiler == "Visual Studio":
-            comp.cxxflags.append("/await")
+            self.cpp_info.cxxflags.append("/await:strict")
```
Comment (hnOsmium0001) [inline recipes/andreasbuhr-cppcoro/all/conanfile.py:102]:
```suggestion
            if tools.Version(self.settings.compiler.version) <= "15"):
                self.cpp_info.cxxflags.append("/await")
            self.cpp_info.cxxflags.append("/await:heapelide")
```

For the first approach I mentioned.

File: recipes/andreasbuhr-cppcoro/all/conanfile.py [hunk 2]
```diff
@@ -91,18 +93,16 @@ def package(self):
...
-        comp.names["cmake_find_package_multi"] = "cppcoro"
-
+        self.cpp_info.libs = ["cppcoro"]
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            self.cpp_info.system_libs.append("pthread")
         if self.settings.compiler == "Visual Studio":
-            comp.cxxflags.append("/await")
+            if tools.Version(self.settings.compiler.version) <= "15"):
+                self.cpp_info.cxxflags.append("/await")
+            self.cpp_info.cxxflags.append("/await:heapelide")
```
Comment (SpaceIm) [inline recipes/andreasbuhr-cppcoro/all/conanfile.py:104]:
```suggestion
            self.cpp_info.cxxflags.append("/await:heapelide")
            if tools.Version(self.settings.compiler.version) < "16":
                self.cpp_info.cxxflags.append("/await")
```

**Comment by ericLemanissier** [comment]
> they may build the libraries with c++20, but compile their own code with c++17

this is not a valid use case, if the recipe does not explicitly say so (https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.htm).  In the general case, conan settings have to be consistent with the final user's settings.

**Pull request closed without merging**